File: 1895-minimum-number-of-operations-to-move-all-balls-to-each-box/1895-minimum-number-of-operations-to-move-all-balls-to-each-box.py
class Solution(object):
    def minOperations(self, boxes):
        """
        :type boxes: str
        :rtype: List[int]
        """
        n = len(boxes)
        res = [0] * n

        balls, moves = 0, 0
        for i in range(n):
            res[i] = (balls + moves)
            moves += balls
            balls += (ord(boxes[i]) - ord('0'))
        
        balls, moves = 0, 0
        for i in reversed(range(n)):
            res[i] += (balls + moves)
            moves += balls
            balls += (ord(boxes[i]) - ord('0'))
        
        return res

File: 1895-minimum-number-of-operations-to-move-all-balls-to-each-box/1895-minimum-number-of-operations-to-move-all-balls-to-each-box.java
class Solution {
    public int[] minOperations(String boxes) {
        int n = boxes.length();
        int b = 0, m = 0;
        int[] res = new int[n];

        for (int i = 0; i < n; i++) {
            res[i] = (b + m);
            m += b;
            b += (boxes.charAt(i) - '0');
        }

        b = m = 0;
        for (int i = n - 1; i >= 0; i--) {
            res[i] += (b + m);
            m += b;
            b += (boxes.charAt(i) - '0');
        }
        return res;
    }
}